src/main/java/frc/robot/utils/MovingAverage.java

```diff
@@ -0,0 +1,114 @@
+package frc.robot.utils;
+
+import java.io.*;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * The moving average contains a list of velocity based on the distance, and should give an approximation of the velocity to apply.
+ * In order to the moving average object to approximate better, you need to pass more values.
+ * The object calculate calculate a straight line between the closest point, and retrieve the velocity (in the straight line between the records).
+ */
+public class MovingAverage {
+    private final String DELIMITER = ",";
+    private final Map<Double, Double> distanceVelocityMap; // <key - distance, value - velocity>, tree map is automatically sort the values by distance.
+
+    public MovingAverage(String pathToCsv) {
+        this(new File(pathToCsv));
+    }
+
+    public MovingAverage(File csv) {
+        this.distanceVelocityMap = new TreeMap<>();
+        if (csv.exists() && csv.isFile()) {
+            try {
+                read(new FileReader(csv));
+            } catch (FileNotFoundException e) { // Can't happen, we have already checked it
+                e.printStackTrace();
+            }
+        } else
+            System.err.println("File not found: " + csv.getName());
+    }
+
+    public MovingAverage(Map<Double, Double> map) {
+        this.distanceVelocityMap = new TreeMap<>(map); // just because I want to order the values by the key.
+    }
+
+    public MovingAverage(double[][] values) {
+        this.distanceVelocityMap = new TreeMap<>();
+        for (double[] pair : values) {
+            if (pair.length >= 2)
+                this.distanceVelocityMap.put(pair[0], pair[1]);
```

Comment:
> What if the excessive value is at the beginning of the row?
What do you mean by that

Reply:
I believe you take `pair[0], pair[1]` so if someone passes an array that looks like that:
| a | b | c |
|---|----|-----|
| 1 | 10 | 100 |
| 2 | 20 | 200 |
| 3 | 30 | 300 |

You  would take from that the following table:
| a | b|
|---|----|
| 1 | 10 |
| 2 | 20 |
| 3 | 30 |

This might not be a correct assumption, maybe I want the values in columns `b` and `c`. That's why I suggested:
> raising an exception is better, or at least print a warning.

Reply:
And why do you think...?